include/envoy/router/scopes.h

```diff
@@ -8,6 +8,77 @@
 namespace Envoy {
 namespace Router {
 
+/**
+ * Scope key fragment base class.
+ */
+class ScopeKeyFragmentBase {
+public:
+  bool operator!=(const ScopeKeyFragmentBase& other) const { return !(*this == other); }
+
+  bool operator==(const ScopeKeyFragmentBase& other) const {
+    if (typeid(*this) == typeid(other)) {
+      return hash() == other.hash();
+    }
+    return false;
+  }
+  virtual ~ScopeKeyFragmentBase() = default;
+
+  // Hash of the fragment.
+  virtual uint64_t hash() const PURE;
+};
+
+/**
+ *  Scope Key is composed of non-null fragments.
+ **/
+class ScopeKey {
+public:
+  ScopeKey() = default;
+  ScopeKey(ScopeKey&& other) = default;
+
+  // Scopekey is not copy-assignable and copy-constructible as it contains unique_ptr inside itself.
+  ScopeKey(const ScopeKey&) = delete;
+  ScopeKey operator=(const ScopeKey&) = delete;
+
+  // Caller should guarantee the fragment is not nullptr.
+  void addFragment(std::unique_ptr<ScopeKeyFragmentBase>&& fragment) {
+    ASSERT(fragment != nullptr, "null fragment not allowed in ScopeKey.");
+    updateHash(*fragment);
+    fragments_.emplace_back(std::move(fragment));
+  }
+
+  uint64_t hash() const { return hash_; }
+  bool operator!=(const ScopeKey& other) const;
+  bool operator==(const ScopeKey& other) const;
+
+private:
+  // Update the key's hash with the new fragment hash.
```

Comment:
why we have these diffs?

Reply:
nvm, I see the move later now.
on the other hand, maybe not move stringkeyfragment here.

Reply:
Arguably we should split this into interface/implementation, since there is a lot of low-level implementation logic here. We have precedence for doing it this way in some cases, e.g. header map, but strongly prefer to not put a ton of implementation logic (e.g. the hash function below) inside `include/`.